### ISSUE TITLE
Refactor REST API Specification (BASIC)

### ISSUE THREAD

**Issue by brian-ruf-ezd** [EasyDynamics/oscal-rest#70]
# User Story
As a developer of OSCAL-enabled tools, I want a REST API specification that - when implemented as a service  - enables me to view and manipulate OSCAL files contained within any organization's repository with this specification implemented.

## Acceptance Criteria

- [ ] The updated REST API specification aligns with the current state of the NIST OSCAL standard
- [ ] The updated REST API specification is published to Swagger for public review and consumption

## Background
With increased understanding of OSCAL within the team, and a more refined understanding of how entities will interact with OSCAL content, we need to revise the REST API specification such that it more completely addresses the user story above.

This includes, but is not limited to:
- better aligning the specification with clarifications made by NIST around OSCAL, such as the use of root-level UUIDs
- moving to an implementation-specific ID assignment specification
  - Possibly including a standard for the ID type, such as a UUIDv4 specification
- Revisiting the data granularity of GET/POST/PATCH/PUT/DELETE operations
  - Consideration for the OSCAL stack as a whole in addition to individual files
  - Complex transactions involving related data in multiple assemblies or even multiple OSCAL files
- Handling of multiple versions of the same OSCAL file.
- Identity management (already included - just better clarify)
- Updating to OpenAPI 3.1

**Comment by brian-ruf-ezd** [comment]
# March 27, 2023 - Team discussion:
- IDs in spec
  - details are implementation defined; however, must meet the following criteria:
    - Uses OSCAL token format: https://pages.nist.gov/OSCAL/reference/datatypes/#token
      - alpha numeric
      - case insensitive
      - only the following special characters: dash, period, underscore
      - no spaces
    - min length: 1
    - max length: 72

**Comment by brian-ruf-ezd** [comment]
# March 27, 2023 - Team Discussion

Clean up spec to clarify the use of oAuth2 tokens within the header of API calls.

NOTE: The implementation will handle the connection between an oAuth2 token and an identity as well as the access rights of that identity.

**Comment by brian-ruf-ezd** [comment]
# From March 22, 2023 - Team Discussion 

Handling of versions:
- /catalogs/{catalogId}/versions?version={versionId}
- /catalogs/{catalogId}/versions?version=latest

**Comment by Bronstrom** [comment]
4/7/23 Sprint Planning - Moved from Sprint 44 to Sprint 45

**Comment by tuckerzp** [comment]
Pushing to Sprint 46

**Comment by brian-ruf-ezd** [comment]
We worked on the following yesterday:

|        | When Valid                                                    | 200 | Return | 201 | Return                  | 400<br>Bad Request | 401<br>Unauthorized | 403<br>Forbidden | 404<br>Not Found | 409<br>Conflict | 410<br>Gone | 415<br>Unsupported Media Type | 422<br>Unprocessable Content |
| ------ | ------------------------------------------------------------- | --- | ------ | --- | ----------------------- | ------------------ | ------------------- | ---------------- | ---------------- | --------------- | ----------- | ----------------------------- | ---------------------------- |
| GET    | Any endpoint                                                  | X   | Object |     |                         |                    | X                   | X                | X                |                 | X           |                               | X                            |
| PUT    | Any endpoint                                                  | X   | Object | X   | Object, Location Header | X                  | X                   | X                | X                | X               | X           | X                             | X                            |
| DELETE | Any endpoint, where cardinality is not exactly 1              | X   | Object |     |                         |                    | X                   | X                | X                | X               | X           |                               |                              |
| PATCH  | Any endpoint that can unambigulously point to a single object | X   | Object |     |                         | X                  | X                   | X                | X                | X               | X           | X                             | X                            |
| POST   | Any endpoint that can unambiguously point to a single object  |     |        | X   | Object, Location Header | X                  | X                   | X                | X                | X               | X           | X                             | X                            |

**Comment by brian-ruf-ezd** [comment]
We updated the above guidance as follows:
|    | **Node Cardinality** | **Node's ID/UUID Cardinality** | **Description Template**                                                                                                                   |
|-------------------------------------------------------|:--------------------:|:------------------------------:|--------------------------------------------------------------------------------------------------------------------------------------------|
|      **GET** by identifier<br>(1 object returned)     |          n/a         |                0               | _[Not possible - do not generate]_                                                                                                         |
|      **GET** by identifier<br>(1 object returned)     |          n/a         |             0 or 1             | When _["an id" \|   "a uuid"]_ is available, returns the _[node name]_ represented by the identifier.                                      |
|      **GET** by identifier<br>(1 object returned)     |          n/a         |                1               | Returns the _[node name]_ represented by the identifier.                                                                                   |
| **GET** no identifier<br>(1 or more objects returned) |          n/a         |               n/a              | Returns every relevant _[node name]_.                                                                                                      |
|                        **PUT**                        |          n/a         |                0               | _[Not possible - do not generate]_                                                                                                         |
|                        **PUT**                        |          n/a         |             0 or 1             | When _["an id" \|   "a uuid"]_ is available, updates the _[node name]_ specified by the identifier, and returns the updated _[node name]_. |
|                        **PUT**                        |          n/a         |                1               | Updates the _[node name]_ specified by the identifier.                                                                                     |
|                       **DELETE**                      |       exactly 1      |               n/a              | _[Not allowed - do not generate]_                                                                                                          |
|                       **DELETE**                      | other than exactly 1 |                0               | _[Not   possible - do not generate]_                                                                                                       |
|                       **DELETE**                      | other than exactly 1 |             0 or 1             | When _["an id" \| "a uuid"]_ is available, deletes the _[node name]_ specified by the identifier.                                          |
|                       **DELETE**                      | other than exactly 1 |                1               | Deletes the _[node name]_ specified by the identifier.                                                                                     |
|                       **PATCH**                       |          n/a         |                0               | _[Not possible - do not generate]_                                                                                                         |
|                       **PATCH**                       |          n/a         |             0 or 1             | When _["an id" \|   "a uuid"]_ is available, updates the _[node name]_ specified by the identifier.                                        |
|                       **PATCH**                       |          n/a         |                1               | Updates the _[node name]_ specified by the identifier.                                                                                     |
|                        **POST**                       |        0 or 1        |               n/a              | When a _[node name]_ does not exist, adds a new _[node name]_.                                                                             |
|                        **POST**                       |       exactly 1      |               n/a              | _[Not allowed - do not generate]_                                                                                                          |
|                        **POST**                       |      0/1 or more     |               n/a              | Adds a new _[node name]_.                                                                                                                  |

**NOTES**:			
- For [node name] use the OSCAL syntax - all lower case - and replace any dash ('-') with a space.			
- For descriptions starting with "When ["an id" | "a uuid"] is available …" fill in only "an id" or only "a uuid" as appropriate.			

and broke the return codes out into a separate table:
|        | **200** | Return | **201** |          Return         | **400**<br>Bad Request | **401**<br>Unauthorized | **403**<br>Forbidden | **404**<br>Not Found | **409**<br>Conflict | **410**<br>Gone | **415**<br>Unsupported Media Type | **422**<br>Unprocessable Content |
|--------|:-------:|:------:|:-------:|:-----------------------:|:----------------------:|:-----------------------:|:--------------------:|:--------------------:|:-------------------:|:---------------:|:---------------------------------:|:--------------------------------:|
| GET    |    X    | Object |         |                         |                        |            X            |           X          |           X          |                     |        X        |                                   |                 X                |
| PUT    |    X    | Object |    X    | Object, Location Header |            X           |            X            |           X          |           X          |          X          |        X        |                 X                 |                 X                |
| DELETE |    X    | Object |         |                         |                        |            X            |           X          |           X          |          X          |        X        |                                   |                                  |
| PATCH  |    X    | Object |         |                         |            X           |            X            |           X          |           X          |          X          |        X        |                 X                 |                 X                |
| POST   |         |        |    X    | Object, Location Header |            X           |            X            |           X          |           X          |          X          |        X        |                 X                 |                 X                |

**Comment by mpemy** [comment]
@brian-easyd  Thanks for posting these updates here.

**Comment by brian-ruf-ezd** [comment]
@mpemy - As discussed, this issue is morphing into "OSCAL REST API Basic" and focuses on file-level operations. A separate issue will be created for "OSCAL REST API Complete" with the more detailed operations.

For every model, we should have:

**PRIMARY**
- `GET /[model-name]`
- `POST /[model-name]`
- `GET /[model-name]/{identifier}`
- `PUT /[model-name]/{identifier}`
- `PATCH /[model-name]/{identifier}`
- `DELETE /[model-name]/{identifier}`


**SNAPSHOTS**
- `GET /[model-name]/{identifier}/snapshot`
- `POST /[model-name]/{identifier}/snapshot` - Creates a new
- `GET /[model-name]/{identifier}/snapshot/{identifier}`
- `PATCH /[model-name]/{identifier}/snapshot/{identifier}` For updating Metadata Properties
- `DELETE /[model-name]/{identifier}/snapshot/{identifier}`
NOTE: Due to the nature of snapshots, there would not be a PUT. We would never replace a snapshot. We might patch it.


**ATTACHMENTS**
- `GET /[model-name]/{identifier}/attachment`
- `POST /[model-name]/{identifier}/attachment`
- `GET /[model-name]/{identifier}/attachment/{identifier}`
- `PUT /[model-name]/{identifier}/attachment/{identifier}`
- `DELETE /[model-name]/{identifier}/attachment/{identifier}`
NOTE: No PATCH as attachments are often binary and patch would be invalid. Attachments should always be replaced, not updated.

**PROFILE RESOLUTION**
For Profiles, we also need to handle resolved profiles. In addition the above, Profile endpoints include:
- `GET /profile/{identifier}/resolved` - returns a resolved profile catalog based on current content
- `POST /profile/{identifier}/resolved-snapshot` - caches a resolved profile catalog at a point in time
- `GET /profile/{identifier}/resolved-snapshot` - retrieves the list of cached resolved profile catalogs
- `GET /profile/{identifier}/resolved-snapshot{identifier}` - retrieves a specific resolved profile catalog snapshot
- `DELETE /profile/{identifier}/resolved-snapshot{identifier}` - removes a cached resolved profile catalog

**Comment by Bronstrom** [comment]
6/8/23 Sprint Planning - Moved from Sprint 48 to Sprint 49

**Comment by Bronstrom** [comment]
6/22/23 Sprint Planning - Moved from Sprint 49 to Sprint 50

**Comment by Bronstrom** [comment]
7/6/23 Sprint Planning - Moved from Sprint 50 to Sprint 51

**Comment by brian-ruf-ezd** [comment]
# July 6 and 10 review and feedback notes

# API Review:

- `GET /*` 
    - return root node should always be "*-list"
        - Example: return root node for `GET /catalog` should be `catalog-list`

- `POST /*` 
    - Should **not** return the full document
    - Instead return the same data as with a single entry in a list
        - Example, `POST /catalog` should return:
        ```
        {
        "file-id": "string",
        "title": "string",
        "published": "2023-07-07T00:36:33.367Z",
        "last-modified": "2023-07-07T00:36:33.367Z",
        "version": "string",
        "oscal-version": "string",
        "document-ids": {
            "document-ids": [
            {
                "document-id": {
                "scheme": "string",
                "identifier": "string"
                }
            }
            ]
        },
        "markings": [
            "string"
        ],
        "remarks": "string"
        }
        ```
    
- `GET /*/{id}` good

- `PUT /*` 
    - No 200 nor 201
    - Use 204 [https://swagger.io/docs/specification/describing-responses/#headers](https://swagger.io/docs/specification/describing-responses/#headers)

- `PATCH /*` good 

- `DELETE /*` good

- `GET /*/{id}/attachment` 
    - return root node should always be "attachment-list"
        - Example: return root node for `GET /catalog/{id}/attachment` should be `attachment-list`
    - return content block should **not** include `base64` field

- `POST /*/{id}/attachment/{id}` good 

- `GET /*/{id}/attachment/{id}` 
    - needs to return the actual file
    
- `PUT /*/{id}/attachment`
    - do **not** include the `base64` field

- `GET /*/{id}/attachment/{id}`
    - See the "Responses That Returns a File" section at [https://swagger.io/docs/specification/describing-responses/](https://swagger.io/docs/specification/describing-responses/)
    - It looks like we'd use one of the following:
    ```
    responses:
        '200':
            description: A binary file in any type
            content:
                multipart/form-data: # Media type
                schema:            # Request payload
                    type: object
                    properties:      # Request parts
                    id:            # Part 1 (string value)
                        type: string
                        format: uuid
                    name:       # Part2 (object)
                        type: string
                    attachment-binary:  # Part 3 (an image)
                        type: string
                        format: binary
    ```


- `PUT /*/{id}/attachment/{id}`
    - No 200 nor 201
    - Use 204 [https://swagger.io/docs/specification/describing-responses/#headers](https://swagger.io/docs/specification/describing-responses/#headers)

- `DELETE /*/{id}/attachment/{id}`
    - No 200
    - Use 204 [https://swagger.io/docs/specification/describing-responses/#headers](https://swagger.io/docs/specification/describing-responses/#headers)


- `GET /*/{id}/snapshot` 
    - return root node should always be "snapshot-list"
        - Example: return root node for `GET /catalog/{id}/snapshot` should be `snapshot-list`

- `POST /*/{id}/snapshot` 
    - should **not** return entire file
    - Instead return the same data as with a single entry in a list
        - Example, `POST /catalog/{id}/shapshot` should return:
        ```
        {
        "file-id": "string",
        "title": "string",
        "published": "2023-07-07T00:36:33.367Z",
        "last-modified": "2023-07-07T00:36:33.367Z",
        "version": "string",
        "oscal-version": "string",
        "document-ids": {
            "document-ids": [
            {
                "document-id": {
                "scheme": "string",
                "identifier": "string"
                }
            }
            ]
        },
        "markings": [
            "string"
        ],
        "remarks": "string"
        }
        ```



- `GET /*/{id}/snapshot/{id}` good

- `PATCH /*/{id}/snapshot/{id}` good

- `DELETE /*/{id}/snapshot/{id}` good

- `GET /profile/{id}/resolved` good

- `GET /profile/{id}/resolved-snapshot` 
    - Do **not** return entire files
    - return root node should always be "resolved-snapshot-list"
    - This should be the list of resolved snapshots:
    ```
    {
        "resolved-snapshot-list": [
            {
            "file-id": "string",
            "title": "string",
            "published": "2023-07-07T01:38:58.912Z",
            "last-modified": "2023-07-07T01:38:58.912Z",
            "version": "string",
            "oscal-version": "string",
            "document-ids": {
                "document-ids": [
                {
                    "document-id": {
                    "scheme": "string",
                    "identifier": "string"
                    }
                }
                ]
            },
            "markings": [
                "string"
            ],
            "remarks": "string"
            }
        ]
    }
    ```

- `POST /profile/{id}/resolved-snapshot` good

- `GET /profile/{id}/resolved-snapshot/{id}` good

- `DELETE /profile/{id}/resolved-snapshot/{id}` good


## Kyle's chat notes:

`PUT` - use 204 No Content [https://swagger.io/docs/specification/describing-responses/#headers](https://swagger.io/docs/specification/describing-responses/#headers)

PATCH - Use `200` and return the document for the "level" where the request was made.

`PATCH /catalog/{catalogId}` should return the full catalog regardless of what fields were included in the response. `PATCH /catalog/metadata/parties/{partyId}` would return the entire modified **party** (note not the catalog) regardless of what field was modified. If the client needs the UUID/last-modified, they can `GET` for the full document

**Brian's note: `PATCH /catalog/metadata/parties/{partyId}` is not part of the basic spec. This will be relevant in the complete spec later.** 

`GET /catalog/{catalogId}/attachment`

This should return an object with:

```
{
  "attachment-list": [
    {
      "id": "string",
      "media-type": "string",
       "name": "string"
    }
  ]
}
```

`GET /catalog/{catalogId}/attachment/{attachmentId}`

- **Header** `Content-Disposition: attachment; filename=NAME` (where NAME is the `name` attribute from the JSON)
- **Header** `Content-Type: TYPE` (TYPE is the `media-type` from the JSON)
- **Body** - The content of the attachment object; this is raw binary data.

We may need to, for OpenAPI reasons, specify the content type as `application/octet-stream` however we should CLEARLY AND THOROUGHLY document that it will be variable based on the specified media type and that `octet-stream` is the FALLBACK

**Comment by brian-ruf-ezd** [comment]
# REST API Review - July 26, 2023

- `POST /[MODEL-NAME]/` return root should be `[MODEL-NAME]-list` instead of simply `[MODEL-NAME]`, which will have exactly one item in it.
    - I'm concerned that simply returning the model name incorrectly signals that the rest of the content is the full OSCAL model rather than just summary details.

- Descriptions in general
    - "ID" should be capitalized in all of the descriptions. ("... by the ID")
    - The description for `attachment` endpoints that end in "... represented by the id" should say "... reprented by the attachment ID" (add "attachment")
    - The description for `snapshot` endpoints that end in "... represented by the id" should say "... reprented by the snapshot ID" (add "snapshot") 
    - The description for `resolved-snapshot` endpoints  that end in "... represented by the id" should say "... reprented by the resolved snapshot ID" (add "resolved snapshot") 

- `PUT /[MODEL-NAME]/{id}` 
    - description should say "Replaces the [model name] represented by the ID." ("Replaces" instead of "Updates")
    - There should be no return data since it's a 204

- `PATCH /[MODEL-NAME]/{id}` description should say "Updates a portion of the [model name] represented by the ID." (inserting "a portion of")

- `PUT /[MODEL-NAME]/{id}/attachment/{id}`
    - description should say "Replaces the attachment represented by the ID." ("Replaces" instead of "Updates")
    - There should be no return data since it's a 204

- DELETE endpoints
  - Including:
    - `DELETE /[MODEL-NAME]/{id}`
    - `DELETE /[MODEL-NAME]/{id}/attachment/{id}`
    - `DELETE /[MODEL-NAME]/{id}/snapshot/{id}`
    - `DELETE /[MODEL-NAME]/{id}/resolved-snapshot/{id}`
  - I've changed my thinking on this. I believe we should simply return a 204 and no data instead of returning the file. 
  - I don't want to transmit a potentially large file back to the client by default - which is especially a risk with attachments. 
  - The developer can always perform a `GET` before a `DELETE` if they want to offer an undo feature.
  
- `POST /[MODEL-NAME]/{id}/snapshot`
    - return root should be `[MODEL-NAME]-list` instead of simply `[MODEL-NAME]`, which will have exactly one item in it.

- `PATCH /[MODEL-NAME]/{id}/snapshot/{id}` description should say "Updates a portion of the snapshot represented by the Snapshot ID." (inserting "a portion of")
